File: js&ts/lc-345.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  const vowels = {
    a: true,
    e: true,
    i: true,
    o: true,
    u: true,
    A: true,
    E: true,
    I: true,
    O: true,
    U: true,
  };
  let pos = [];
  for (let i = 0; i < length(s); i++) {
    if (vowels[s[i]] === true) {
      pos.push(i);
    }
  }
  let i = 0;
  let j = pos.length - 1;
  while (i < j) {
    tmp = s[pos[i]];
    s[pos[i]] = s[pos[j]];
    s[pos[j]] = tmp;
  }
  return s;
};

File: lc-345.go
package main

func reverseVowels(s string) string {
	vowels := map[byte]bool{
		'a': true,
		'e': true,
		'i': true,
		'o': true,
		'u': true,
		'A': true,
		'E': true,
		'I': true,
		'O': true,
		'U': true,
	}
	pos := []int{}
	for ind, _ := range s {
		if vowels[s[ind]] == true {
			pos = append(pos, ind)
		}
	}
	i, j := 0, len(pos)-1
	res := []byte(s)
	for i < j {
		res[pos[i]], res[pos[j]] = res[pos[j]], res[pos[i]]
		i++
		j--
	}
	return string(res)
}
